pkg/compact/compact.go

```diff
@@ -309,18 +342,81 @@ func (g *DefaultGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (res []*Gro
 			return nil, errors.Wrap(err, "add compaction group")
 		}
 	}
+
+	if g.planner != nil {
+		newGroups := []*Group{}
+		// split each group by available plans
+		for _, group := range res {
+			newGroup, err := g.splitGroupByPlans(group)
+			if err != nil {
+				return nil, err
+			}
+			newGroups = append(newGroups, newGroup...)
+		}
+		res = newGroups
+	}
+
 	sort.Slice(res, func(i, j int) bool {
 		return res[i].Key() < res[j].Key()
 	})
 	return res, nil
 }
 
+// Recursively plan a groups metas until there are no available plans and return a new group for each plan
+// This allows a single group key to utilize group concurrency when there are multiple plans available to compact
+func (g *DefaultGrouper) splitGroupByPlans(group *Group) ([]*Group, error) {
+	groups := []*Group{}
+	// split each group by available plans
+	var seededRand = rand.New(rand.NewSource(time.Now().UnixNano()))
+	for toCompact, err := g.planner.Plan(context.Background(), group.metasByMinTime); len(toCompact) != 0; toCompact, err = g.planner.Plan(context.Background(), group.metasByMinTime) {
+		if err != nil {
+			return nil, errors.Wrap(err, "split group")
+		}
+		lbls := group.Labels()
+		hash := strconv.Itoa(seededRand.Int())
+		groupKey := group.Key()
+		newGroup, err := NewGroup(
+			log.With(g.logger, "split group", fmt.Sprintf("%d@%v", group.Resolution(), lbls.String()), "groupKey", groupKey, "hash", hash),
+			g.bkt,
+			groupKey,
+			hash,
+			lbls,
+			group.Resolution(),
+			group.acceptMalformedIndex,
+			g.enableVerticalCompaction,
+			g.compactions.WithLabelValues(groupKey),
+			g.compactionRunsStarted.WithLabelValues(groupKey),
+			g.compactionRunsCompleted.WithLabelValues(groupKey),
+			g.compactionFailures.WithLabelValues(groupKey),
+			g.verticalCompactions.WithLabelValues(groupKey),
+			g.garbageCollectedBlocks,
+			g.blocksMarkedForDeletion,
+		)
+		if err != nil {
+			return nil, errors.Wrap(err, "create split compaction group")
+		}
+		groups = append(groups, newGroup)
+		for _, m := range toCompact {
+			if err := newGroup.Add(m); err != nil {
+				return nil, errors.Wrap(err, "add to split compaction group")
+			}
+			for i, mm := range group.metasByMinTime {
+				if mm.ULID == m.ULID {
+					group.metasByMinTime = append(group.metasByMinTime[:i], group.metasByMinTime[i+1:]...)
```

Comment:
I wonder if this work well. 

1. First of all it does not detect places where output block has to be used for next compaction, that results in unnecessary compaction without output block causing vertical compaction later on.
2.  This still does not detect  case mentioned in issue which is overlapped blocks not being grouped into single compact, no?

Reply:
1. Again this feature wasn't meant to deduce what can be used for next compactions. If thats desired, I'll have to put some thought into that. 
2. Im not sure what you mean by that. This line of code is to remove metas from the parent group that have already been planned into a new group. The case mentioned in the issue is that you are behind and have multiple windows of overlapping blocks that can be compacted concurrently.

Reply:
I will create a separate issue to handle predicting what blocks will be needed next and build on this work.